### ISSUE TITLE
[BUG] Disk pressure caused by Rancher agent tmp files

### ISSUE THREAD

**Issue by jcambronero** [harvester/harvester#2821]
**Describe the bug**
After a few weeks of running Harvester we started getting Kubelet disk pressure errors.
When looking further we noticed that there were >50GBs generated by thousands of tmp directories in the /usr/local/.state/var-lib-rancher.bind/agent/tmp directory
![image](https://user-images.githubusercontent.com/6895571/191878685-018304c1-b630-4370-be9f-ecfd6df6288e.png)

**To Reproduce**
Run Harvester for a few days, you'll start noticing several tmp directories created in /usr/local/.state/var-lib-rancher.bind/agent/tmp on each of the Harverster nodes.

**Expected behavior**
These tmp directories should be cleaned up automatically.

**Environment**
 - Harvester ISO version: v1.0.0
 - Underlying Infrastructure: Baremetal

**Comment by bk201** [comment]
Hi @jcambronero, any notable logs in `journalctl -u rancher-system-agent.service`? Thanks.

**Comment by jcambronero** [comment]
hi @bk201, I'm not seeing anything interesting. These logs are repeating every few seconds.
BTW, I deleted all the tmp directories yesterday and today there's already 36 new tmp directories (and growing).

Here's the logs:
```
Sep 23 15:59:29 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:29Z" level=debug msg="[K8s] Enqueueing after 5.000000 seconds"
Sep 23 15:59:29 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:29Z" level=debug msg="[K8s] secret data/string-data did not change, not updating secret"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] Processing secret custom-605ce9580dc9-machine-plan in namespace fleet-local at generation 0 with resource version 12591712"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] Calculated checksum to be ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] Remote plan had an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] Applied checksum was the same as the plan from remote. Not applying."
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] last applied resource version (12591712) did not change. running probes, skipping apply."
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] needsApplied was false, not applying"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] writing an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0 to the remote plan"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (kubelet) running probe"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (kubelet) retrieving existing probe status from map if existing"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (calico) running probe"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (calico) retrieving existing probe status from map if existing"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe output was ok"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe output was "
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe succeeded"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (calico) writing probe status to map"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="Probe succeeded"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[Prober] (kubelet) writing probe status to map"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] Enqueueing after 5.000000 seconds"
Sep 23 15:59:34 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:34Z" level=debug msg="[K8s] secret data/string-data did not change, not updating secret"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] Processing secret custom-605ce9580dc9-machine-plan in namespace fleet-local at generation 0 with resource version 12591712"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] Calculated checksum to be ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] Remote plan had an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] Applied checksum was the same as the plan from remote. Not applying."
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] last applied resource version (12591712) did not change. running probes, skipping apply."
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] needsApplied was false, not applying"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] writing an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0 to the remote plan"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (kubelet) running probe"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (kubelet) retrieving existing probe status from map if existing"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (calico) running probe"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (calico) retrieving existing probe status from map if existing"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe output was "
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe succeeded"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (calico) writing probe status to map"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe output was ok"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="Probe succeeded"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[Prober] (kubelet) writing probe status to map"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] Enqueueing after 5.000000 seconds"
Sep 23 15:59:39 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:39Z" level=debug msg="[K8s] secret data/string-data did not change, not updating secret"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] Processing secret custom-605ce9580dc9-machine-plan in namespace fleet-local at generation 0 with resource version 12591712"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] Calculated checksum to be ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] Remote plan had an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] Applied checksum was the same as the plan from remote. Not applying."
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] last applied resource version (12591712) did not change. running probes, skipping apply."
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] needsApplied was false, not applying"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] writing an applied checksum value of ff4437f55b61b745a56e583d9aa9259c4d50c0a6e2bd5bfd89ac44380f225bc0 to the remote plan"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (kubelet) running probe"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (kubelet) retrieving existing probe status from map if existing"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (calico) running probe"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (calico) retrieving existing probe status from map if existing"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe timeout duration: 5 seconds"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe output was ok"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe succeeded"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (kubelet) writing probe status to map"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe output was "
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Setting success threshold to 1"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Setting failure threshold to 2"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="Probe succeeded"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[Prober] (calico) writing probe status to map"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] Enqueueing after 5.000000 seconds"
Sep 23 15:59:44 <hostname> rancher-system-agent[1979]: time="2022-09-23T15:59:44Z" level=debug msg="[K8s] secret data/string-data did not change, not updating secret"
```

**Comment by w13915984028** [comment]
@jcambronero Is your `Harvester` cluster managed by an `Rancher`, any abnormal inbetween them ?

**Comment by jcambronero** [comment]
Hi @w13915984028, this Harvester cluster is managed by Rancher. 
I'm not seeing any immediate issues there. The cluster looks active. Anything specific I should be looking for?

**Comment by rodolfoalmeid** [comment]
Hello,
We are facing the same issue in a 3 nodes cluster.
Harvester 1.0.3
One of the nodes is showing a message: "Kubelet has disk pressure: Resource is Ready"
That failed node has  /usr/local/ 87% full.

harvester-01:~> df -h /usr/local
Filesystem Size Used Avail Use% Mounted on /dev/sdg5 98G 81G 13G 87% /usr/local

harvester-01:/ # du -h --max-depth=1 /usr/local/.state/var-lib-rancher.bind/agent/tmp
29M     /usr/local/.state/var-lib-rancher.bind/agent/tmp/tmp.1ew9mVqhrU
29M     /usr/local/.state/var-lib-rancher.bind/agent/tmp/tmp.YHuQtpUWJr
29M     /usr/local/.state/var-lib-rancher.bind/agent/tmp/tmp.qvqIEhsgkg
29M     /usr/local/.state/var-lib-rancher.bind/agent/tmp/tmp.F8P3HksiTY
….etc

In summary /usr/loca/.state/var-lib-rancher.bind/agent/tmp has 73GB

**Comment by bk201** [comment]
@FrankYang0529 Please take a look. Thanks!

**Comment by FrankYang0529** [comment]
I tried to run a harvester v1.0.3 cluster for two days. However, there is no any folder under `/usr/local/.state/var-lib-rancher.bind/agent/tmp`. I only found a script [`run.sh`](https://github.com/rancher/system-agent/blob/00181cd06b60990ab12347c7826de951d54ab7d9/package/suc/run.sh#L5-L9) creates temporary folders, so I created https://github.com/rancher/system-agent/pull/100 to clean up temporary folders even if the script is failed.

**Comment by FrankYang0529** [comment]
Hello @jcambronero and @rodolfoalmeid, do you still have the cluster? If yes, could you generate the support bundle and upload it to GitHub? Also, could you print the folder structure under `tmp.xxxx` folder? Thank you!

**Comment by jcambronero** [comment]
Hi @FrankYang0529, I think we found the source of the tmp directories.

We noticed that in the cattle-system events there's a loop of `apply-system-agent-upgrade-on-<hostname>` pods being continuously created/deleted.

The log of these pods looks like this:
```
+ TMPDIRBASE=/var/lib/rancher/agent/tmp
+ mkdir -p /host/var/lib/rancher/agent/tmp
+ chroot /host /bin/sh -c 'mktemp -d -p /var/lib/rancher/agent/tmp'
+ TMPDIR=/var/lib/rancher/agent/tmp/tmp.hISItxuVyd
+ cp /opt/rancher-system-agent-suc/install.sh /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd
+ cp /opt/rancher-system-agent-suc/rancher-system-agent /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd
+ chmod +x /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/install.sh
+ '[' -n <hostname> ]
+ NODE_FILE=/host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/node.yaml
+ kubectl get node <hostname> -o yaml
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/etcd: "true"' /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/node.yaml
+ export 'CATTLE_ROLE_ETCD=true'
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/controlplane: "true"' /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/node.yaml
+ '[' -z  ]
+ grep -q 'node-role.kubernetes.io/control-plane: "true"' /host/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/node.yaml
+ export 'CATTLE_ROLE_CONTROLPLANE=true'
+ '[' -z true ]
+ export 'CATTLE_AGENT_BINARY_LOCAL=true'
+ export 'CATTLE_AGENT_BINARY_LOCAL_LOCATION=/var/lib/rancher/agent/tmp/tmp.hISItxuVyd/rancher-system-agent'
+ '[' -s /host/etc/systemd/system/rancher-system-agent.env ]
+ grep -v '^#' /host/etc/systemd/system/rancher-system-agent.env
+ var=HTTP_PROXY
+ val=
+ eval 'v="$HTTP_PROXY"'
+ v=http://<proxy-host>:<port>
+ '[' -z http://<proxy-host>:<port> ]
+ var=HTTPS_PROXY
+ val=
+ eval 'v="$HTTPS_PROXY"'
+ v=http://<proxy-host>:<port>
+ '[' -z http://<proxy-host>:<port> ]
+ var=NO_PROXY
+ val=,localhost,127.0.0.1,0.0.0.0,10.0.0.0/8,cattle-system.svc,.svc,.cluster.local
+ eval 'v="$NO_PROXY"'
+ v=localhost,127.0.0.1,127.0.0.0/8,10.0.0.0/8,cattle-system.svc,.cattle-system.svc,.svc,.cluster.local,harvester-system,cattle-system,longhorn-system,.harvester-system,.cattle-system,.longhorn-system,0.0.0.0
+ '[' -z localhost,127.0.0.1,127.0.0.0/8,10.0.0.0/8,cattle-system.svc,.cattle-system.svc,.svc,.cluster.local,harvester-system,cattle-system,longhorn-system,.harvester-system,.cattle-system,.longhorn-system,0.0.0.0 ]
+ chroot /host /var/lib/rancher/agent/tmp/tmp.hISItxuVyd/install.sh
[INFO]  Using default agent configuration directory /etc/rancher/agent
[INFO]  Using default agent var directory /var/lib/rancher/agent
[WARN]  /usr/local is read-only or a mount point; installing to /opt/rancher-system-agent
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
[ERROR]  Error while connecting to Rancher to verify CA necessity. Sleeping for 5 seconds and trying again.
```

The log of the system-upgrade-controller pod look like like this:
```
time="2022-11-22T01:17:52Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-1fa84': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-1fa84\" not found, requeuing"
time="2022-11-22T01:17:54Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-d6452': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-d6452\" not found, requeuing"
time="2022-11-22T01:17:54Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-151aa': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-151aa\" not found, requeuing"
time="2022-11-22T01:17:57Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-20587': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-20587\" not found, requeuing"
time="2022-11-22T01:17:59Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-a966a': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-a966a\" not found, requeuing"
time="2022-11-22T01:18:00Z" level=error msg="error syncing 'cattle-system/apply-system-agent-upgrader-on-<hostname>-with-31ee2-901e6': handler system-upgrade-controller: jobs.batch \"apply-system-agent-upgrader-on-<hostname>-with-31ee2-901e6\" not found, requeuing"
```

The folder structure for the tmp directories looks like this:
![image](https://user-images.githubusercontent.com/93346099/203200342-09916212-b79f-4c41-a8e9-22b8bf99a913.png)


It looks like every new  `apply-system-agent-upgrade-<hostname>` creates a new tmp directory and it's doing this over and over until the disk fills up.
I think we should prevent this behavior **even** if the pods fail.

Also, any ideas on what could be causing the `Error while connecting to Rancher to verify CA necessity` errors?

Thanks!

**Comment by FrankYang0529** [comment]
Hi @jcambronero, thanks for your information! The `tmp.xxx` folder is from [run.sh](https://github.com/rancher/system-agent/blob/v0.2.3/package/suc/run.sh) and the CA error is from [install.sh](https://github.com/rancher/system-agent/blob/00181cd06b60990ab12347c7826de951d54ab7d9/install.sh#L641-L682). We've already merged https://github.com/rancher/system-agent/pull/100 to remove `tmp.xxx` folder when the script exits, so you will not have remaining folders in the next version.

However, I don't have any idea about the CA error. Could you try to run the following commands in any harvester node and see the result? If you get a result `0` or `60`, it should breaks the while loop in install.sh. Thank you.

```sh
# get rancher service cluster-IP
$ kubectl get service -n cattle-system rancher

$ curl --noproxy '*' --connect-timeout 5 --max-time 5 -v "https://<rancher service cluster-IP>/healthz"

$ echo "$?"
```

If you don't have any data on the current cluster, I recommend you try a newer version like [v1.1.0](https://github.com/harvester/harvester/releases/tag/v1.1.0). If you have data, you can try to backup VMs to NFS/S3 and [restore VMs on a new cluster](https://docs.harvesterhci.io/v1.1/vm/backup-restore#restore-a-new-vm-on-another-harvester-cluster).

**Comment by jcambronero** [comment]
Thank you @FrankYang0529, this is what I see:

```
$ kubectl get service -n cattle-system rancher
NAME      TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
rancher   ClusterIP   10.53.196.154   <none>        80/TCP,443/TCP   123d

$ curl --noproxy '*' --connect-timeout 5 --max-time 5 -v "https://10.53.196.154/healthz"
*   Trying 10.53.196.154:443...
* TCP_NODELAY set
* Connected to 10.53.196.154 (10.53.196.154) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (OUT), TLS alert, unknown CA (560):
* SSL certificate problem: unable to get local issuer certificate
* Closing connection 0
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.

$ echo "$?"
60
```

Here's the certificate details:
```
$ echo | openssl s_client -showcerts -servername 10.53.196.154 -connect 10.53.196.154:443 2>/dev/null | openssl x509 -inform pem -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 4317132674176367447 (0x3be9894b4483bb57)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: O = dynamiclistener-org, CN = dynamiclistener-ca
        Validity
            Not Before: Jul 29 23:03:09 2022 GMT
            Not After : Nov  9 05:14:01 2023 GMT
        Subject: O = dynamic, CN = dynamic
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:27:e9:35:32:77:33:5d:a9:6d:0d:f8:46:c3:35:
                    37:85:1e:b3:e5:24:e4:80:64:13:8d:25:8d:b6:29:
                    42:87:40:5b:94:c8:3b:4f:e3:df:eb:3c:e0:96:75:
                    60:1d:4a:83:3f:76:b5:b8:e2:a5:9b:66:bf:31:66:
                    28:3f:f8:25:97
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication
            X509v3 Authority Key Identifier:
                keyid:AF:7C:DA:73:87:1C:4B:C2:2F:75:47:53:88:6F:80:B6:6A:34:75:8D

            X509v3 Subject Alternative Name:
                DNS:localhost.localdomain, DNS:<myhost-1>, DNS:<myhost-2>, IP Address:<my-ip-1>, IP Address:<my-ip-2>, IP Address:<my-ip-3>, IP Address:<my-ip-4>, IP Address:<my-ip-5>, IP Address:<my-ip-6>, IP Address:<my-ip-7>, IP Address:<my-ip-8>, IP Address:<my-ip-9>
    Signature Algorithm: ecdsa-with-SHA256
         30:45:02:20:37:2c:cf:29:e6:5a:72:95:e3:7c:5b:3d:db:33:
         2b:3e:ee:3b:fe:87:f3:b4:84:c6:1c:22:ca:aa:d5:88:c7:52:
         02:21:00:81:dd:b1:8b:09:47:78:fc:8e:a8:c1:1f:8c:69:a3:
         27:af:84:f2:8e:3b:a1:d0:72:72:3d:6c:8e:af:95:ce:17
```

We will try to upgrade to v1.1.0 as suggested. 
Any workarounds that you can think of in the meantime?

Thanks,
Jose

**Comment by janeczku** [comment]
Observed this behaviour in v1.1.0: The directory `/usr/local/.state/var-lib-rancher.bind/agent/tmp` at some point consumed 90% of disk space of the underlying file system. This caused a node pressure eviction resulting in the termination of the Harvester system pod on multiple nodes.

After freeing up disk space, the taint was not automatically removed. Manually removing the taint resulted in the following error: `failed to call webhook: Post [https://harvester-webhook.harvester-system.svc:443/v1/webhook/validation?timeout=10s](https://harvester-webhook.harvester-system.svc/v1/webhook/validation?timeout=10s): no endpoints available for service "harvester-webhook"`

Additional Info:

![image](https://user-images.githubusercontent.com/3813921/212279920-b12d265c-6394-4601-932c-d5240798467e.png)


```
kubectl describe pod harvester-64c68d49f-zxkfh -n harvester-system


Events:

  Type     Reason            Age                  From               Message

  ----     ------            ----                 ----               -------

  Warning  FailedScheduling  59s (x27 over 131m)  default-scheduler  0/4 nodes are available: 4 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }. preemption: 0/4 nodes are available: 4 Preemption is not helpful for scheduling.
Events:

```
![image](https://user-images.githubusercontent.com/3813921/212279158-28510a42-676a-4686-8719-987332346659.png)

**Comment by FrankYang0529** [comment]
> After freeing up disk space, the taint was not automatically removed. Manually removing the taint resulted in the following error: `failed to call webhook: Post [https://harvester-webhook.harvester-system.svc:443/v1/webhook/validation?timeout=10s](https://harvester-webhook.harvester-system.svc/v1/webhook/validation?timeout=10s): no endpoints available for service "harvester-webhook"`

Hi @janeczku, this looks like a dead lock. Nodes have taints, so pods can't be scheduled. However, we also have a validation webhook, so removing taint request should be processed by webhook. I tried following test steps, we can fix this by removing `Validatingwebhookconfigurations`.

1. Add `NoSchedule` taints on all nodes.
2. Remove all harvester-webhook pods. Now, new harvester-webhook pods can't be scheduled.
3. Remove `harvester-validator` Validatingwebhookconfigurations.
4. Remove taints on all nodes. Now, new harvester-webhhok pods can be scheduled.

**Comment by janeczku** [comment]
@FrankYang0529 Thanks for providing the workaround. Do we know what causes the disk filesystem to fill up over time under `/usr/local/.state/var-lib-rancher.bind/agent/tmp` ?

**Comment by w13915984028** [comment]
@janeczku Shortly speaking, in some cases (e.g. CA invalid) the Rancher bash script returns/times out before it deletes the temp file, the fix adds a hook to make sure to delete the tmp file in any return path, avoid it is left there.

**Comment by w13915984028** [comment]
@FrankYang0529 @bk201 @guangbochen 

Regarding #2821 and https://github.com/harvester/harvester/issues/2955, it is essential to add the workaround (https://github.com/harvester/harvester/issues/2821#issuecomment-1383429966)  into trouble-shooting, for manual recover.

**Comment by bk201** [comment]
To bump Rancher to v2.7.x in 1.2.0 (currently v2.6.11-rc11 still doesn't contain updated system-agent image)

**Comment by lanfon72** [comment]
Closed as verified by checking `Rancher System Agent version` been upgrade to `v0.3.2`

![image](https://github.com/harvester/harvester/assets/5169694/887e3d75-5457-4974-8101-5940bb658144)


Test Information
----
* Environment: **qemu/KVM 2 nodes**
* Harvester Version: **v1.2-a5e74302-head**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Login to console, execute `journalctl -u rancher-system-agent.service | grep starting | tail -1`
3. **Rancher System Agent version** should `>= v0.3.0`
